crates/bevy_ui/src/interaction_states.rs

```diff
@@ -0,0 +1,108 @@
+/// This module contains components that are used to track the interaction state of UI widgets.
+use bevy_a11y::AccessibilityNode;
+use bevy_ecs::{
+    component::Component,
+    observer::Trigger,
+    world::{DeferredWorld, OnAdd, OnRemove, OnReplace},
+};
+
+/// A component indicating that a widget is disabled and should be "grayed out".
+/// This is used to prevent user interaction with the widget. It should not, however, prevent
+/// the widget from being updated or rendered, or from acquiring keyboard focus.
+///
+/// For apps which support a11y: if a widget (such as a slider) contains multiple entities,
+/// the `InteractionDisabled` component should be added to the root entity of the widget - the
+/// same entity that contains the `AccessibilityNode` component. This will ensure that
+/// the a11y tree is updated correctly.
+#[derive(Component, Debug, Clone, Copy, Default)]
+pub struct InteractionDisabled(pub bool);
+
+impl InteractionDisabled {
+    /// Returns whether the widget is currently disabled.
+    pub fn get(&self) -> bool {
+        self.0
+    }
+
+    /// Sets the disabled state of the widget.
+    pub fn set_disabled(&mut self) {
+        self.0 = true;
+    }
+
+    /// Clears the disabled state of the widget.
+    pub fn clear_disabled(&mut self) {
+        self.0 = false;
+    }
+}
```

Comment:
Isn't the inner field already public?

Reply:
Yeah this naming feels a bit awkward. Its worth considering either relying on `interaction_disabled.0`, or reducing the stutter on `interaction_disabled.set_disabled()`

Reply:
I think we still need to resolve the question of whether `InteractionDisabled` should be an immutable component or not - because if it's immutable, then all these methods will need to get removed anyway.

The reason why I didn't make it immutable in the first place is that, unlike the other boolean components in the PR, this one is only read by the framework rather than being mutated by it - the the cost/benefit of immutability looks different depending on whether you are a producer or consumer of a component (it's generally great for consumers but inconvenient for producers). However, I'm on the fence here because the common idioms and best practices around Bevy immutability haven't really settled yet.